Hotsapi.Uploader.Common/Manager.cs

```diff
@@ -115,39 +118,76 @@ public async void Start()
 
             _analyzer.MinimumBuild = await _uploader.GetMinimumBuild();
 
-            for (int i = 0; i < MaxThreads; i++) {
-                Task.Run(UploadLoop).Forget();
-            }
+            _ = Task.Run(() => ParseLoop());
+            _ = Task.Run(() => FingerprintLoop());
+            _ = Task.Run(() => UploadLoop());
         }
 
-        private async Task UploadLoop()
-        {
-            while (true) {
-                try {
-                    var file = await processingQueue.TakeAsync();
-
-                    file.UploadStatus = UploadStatus.InProgress;
-
-                    // test if replay is eligible for upload (not AI, PTR, Custom, etc)
-                    var replay = _analyzer.Analyze(file);
-                    if (file.UploadStatus == UploadStatus.InProgress) {
-                        // if it is, upload it
-                        await _uploader.Upload(file);
-                    }
-                    SaveReplayList();
-                    if (ShouldDelete(file, replay)) {
-                        DeleteReplay(file);
-                    }
-                }
-                catch (Exception ex) {
-                    _log.Error(ex, "Error in upload loop");
-                }
+        private async Task ParseLoop() {
+            using (var rateLimitParsing = new SemaphoreSlim(MaxThreads)) {
+                while (await processingQueue.OutputAvailableAsync()) {
+                    try {
+                        var file = await processingQueue.TakeAsync();
+                        await rateLimitParsing.Locked(() => {
+                            _ = WorkerPool.RunBackground(async () => {

```

Comment:
This is messed up, the parse pool is limited  by the background pool. This shouldn't be protected by the semaphore at all. Good catch!

Reply:
Ah okay so it was working as it should, I just got stuck on the wrong bit

Reply:
Ok so you are saying the 'while (await processingQueue.OutputAvailableAsync())' will always return true, it will just wait until something is available for processing

Reply:
It will return true untill `processingQueue.doneAdding()` is called, which signals that there will be no more entries. That doesn't happen in this codebase, so so far, it's the equivalent of `while (true)` It will finish if something every does call that. It could be part of a graceful shutdown procedure for example.